File: 2021/09/14/main.go
package main

func reverseOnlyLetters(s string) string {
	bs := []byte(s)

	l, r := 0, len(bs)-1
	for l < r {
		for l < r && !isLetter(bs[l]) {
			l++
		}
		for l < r && !isLetter(bs[r]) {
			r--
		}
		bs[l], bs[r] = bs[r], bs[l]
		l++
		r--
	}

	return string(bs)
}

func isLetter(c byte) bool {
	return ('a' <= c && c <= 'z') || ('A' <= c && c <= 'Z')
}

File: 2021/09/14/main.py
class Solution:
    def reverseOnlyLetters(self, s: str) -> str:
        l, r = 0, len(s) - 1
        b = list(s)
        def isLetter(c):
            return ('a' <= c and c <= 'z') or ('A' <= c and c <= 'Z')
        while l < r:
            while l < r and not isLetter(b[l]):
                l += 1
            
            while l < r and not isLetter(b[r]):
                r -= 1
            
            b[l], b[r] = b[r], b[l]
            l += 1
            r -= 1
    
        return ''.join(b)

